Allow specifying allowed classes for deserialization

artem-smotrakov opened the pull request:
This patch fixes #1783

1. Moved `SaferObjectInputStream` to `org.springframework.security.oauth2.common.util`. The class allow specifying custom list of allowed classes.
2. Added `SerializationStrategy` based on `SerializationUtils`. The new class uses `SaferObjectInputStream` and defines the default list of allowed classes for deserialization. The class allow specifying custom list of allowed classes.
3. Revert changes in `SerializationUtils` and marked it as deprecated.
4. Updated JDBC services to use `SerializationStrategy` instead of `SerializationUtils`. The classes allow specifying custom `SerializationStrategy`.
5. Updated `JdkSerializationStrategy` (used with Redis)` to use `SerializationStrategy`. The class allow specifying custom list of allowed classes.

artem-smotrakov commented:
@vaddemgen Could you please take a look at the patch?

vaddemgen commented:
@artem-smotrakov **I have just checked and it works. **

**_But I have a question about using this deserialization mechanism_**.

Could you check if packages are properly added to the list of allowed during deserialization? I think, that was the point.

I had to add changes to the code. Tthe other implementation of ```TokenStore``` (`JdbcTokenStore`) was used. Here, during the initialization of the TokenStore bean, an instance of SerializationStrategy was created with my list of allowed classes:

```java
package my.project.config.oauth2;

import my.project.model.CustomPersonDetails;
import org.springframework.security.oauth2.common.util.SerializationStrategy;
import org.springframework.security.oauth2.provider.token.store.JdbcTokenStore;
// ...

@Configuration
@EnableAuthorizationServer
public class AuthorizationServerConfiguration extends AuthorizationServerConfigurerAdapter {

  // ...

  @Bean
  public TokenStore tokenStore() {
    // I use JdbcTokenStore instead of the default InMemoryTokenStore
    JdbcTokenStore jdbcTokenStore = new JdbcTokenStore(dataSource);
    jdbcTokenStore.setSerializationStrategy(new SerializationStrategy(
        List.of(
            // Here I add the packages provided by the constant 
            // SerializationStrategy#DEFAULT_ALLOWED_CLASSES
            "java.lang.",
            "java.util.",
            "org.springframework.security.",
            // And I add my own implementation of PersonDetails 
            // to the list of allowed packages during the deserialization
            CustomPersonDetails.class.getPackageName()
        )
    ));
    return jdbcTokenStore;
  }

  // ...
}
```

It seems to me it's better to also add the ability to customize this list by an env variable.

artem-smotrakov commented:
@vaddemgen Thanks for the feedback and testing the patch.

> Could you check please if packages are properly added to the list of allowed during deserialization? 

Your code looks correct to me. Do you mean the default list of allowed classes should be extended?

> It seems to me it's better to also add the ability to customize this list by an env variable.

I think it may be a nice improvement. Alternatively, we can introduce a system property to customize the list.

@jgrandja What do you think?

artem-smotrakov commented:
If we have a system property or an environment variable, then application code may not be updated which is good.

vaddemgen commented:
@artem-smotrakov 

> Do you mean the default list of allowed classes should be extended?

No, It looks good. Thank you for the answer!

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SaferObjectInputStream.java:
```diff
@@ -0,0 +1,102 @@
+ *
+ * @author Artem Smotrakov
+ */
+public class SaferObjectInputStream extends ObjectInputStream {
```
Let's rename this to `WhitelistedObjectInputStream`.

Can we also make this a `private class` internal (inner) to `SerializationStrategy`.

artem-smotrakov replied:
Sure, no problem. However, it can't be a private class since `SerializationStrategy` is now an interface.

artem-smotrakov replied:
I'll restrict the access to package-private.

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SerializationStrategy.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Serialization strategy based on {@link SaferObjectInputStream}.
+ */
+public class SerializationStrategy {
```
Let's make this an `interface` and name the whitelisted implementation `WhitelistedSerializationStrategy` or `DefaultSerializationStrategy`

artem-smotrakov replied:
Okay, no problem.

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SerializationUtils.java:
```diff
@@ -8,6 +8,10 @@
+/**
+ * This class should not be used since it uses unsafe deserialization via {@link ConfigurableObjectInputStream}.
+ */
+@Deprecated
```
`SerializationUtils` is still used in certain places. We should ensure there is no further usages since `SerializationStrategy` is now preferred.

artem-smotrakov replied:
Right, I missed several places where it's used. I'll replace `SerializationUtils` with `WhitelistedSerializationStrategy`.

I also noticed that some tests use `org.springframework.util.SerializationUtils`. I guess it can be also replaced with `WhitelistedSerializationStrategy`.

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/token/store/redis/JdkSerializationStrategy.java:
```diff
@@ -1,26 +1,89 @@
+   *
+   * @param allowedClasses The allowed classes.
+   */
+  public JdkSerializationStrategy(List<String> allowedClasses) {
```
Please remove this constructor and add `setSerializationStrategy(SerializationStrategy serializationStrategy)`, now that `SerializationStrategy`  is an interface

artem-smotrakov replied:
Okay, I can do that. I thought it might be better to make `serializationStrategy` filed final.

artem-smotrakov commented:
Thanks for the review @jgrandja I have addressed the comments, please take a look.

We can also introduce a way how a user can set a list of allowed classes via environment variable or system property (please see the comment above from @vaddemgen). What do you think?

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SerializationStrategy.java:
```diff
@@ -0,0 +1,99 @@
+   * Special ObjectInputStream subclass that checks if classes are allowed to deserialize.
+   * The class should be configured with a whitelist of only allowed (safe) classes to deserialize.
+   */
+  class WhitelistedObjectInputStream extends ObjectInputStream {
```
Let's move this to `WhitelistedSerializationStrategy` as a `private class` since it's part of that implementation.

artem-smotrakov replied:
Okay, no problem.

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SerializationStrategy.java:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Defines how objects are serialized an deserialized.
+ *
+ * @author Artem Smotrakov
```
Add `@since 2.4`

artem-smotrakov replied:
Sure.

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SerializationUtils.java:
```diff
@@ -8,6 +8,10 @@
 
 import org.springframework.core.ConfigurableObjectInputStream;
 
+@Deprecated
```
I'm debating whether we should deprecate. I feel there are a lot of touch points with `SerializationStrategy` via `setSerializationStrategy()` in the various collaborators. My preference would be to touch less of the code. Since all collaborators use `SerializationUtils` it might be better to isolate the changes there.

We could add a `SerializationStrategy` static member and default to `WhitelistedSerializationStrategy`. The application could override the behaviour via these options:

1. Add new `public static void setSerializationStrategy(SerializationStrategy serializationStrategy)`
2. Using `spring.factories`

With `spring.factories`, the user could override by adding a classpath resource under `META-INF/spring.factories` with the contents being:

`org.springframework.security.oauth2.common.util.SerializationStrategy = mypackage.CustomSerializationStrategy`

You can use `SpringFactoriesLoader` to check if the user has supplied a custom `SerializationStrategy`.

What are your thoughts on this?

artem-smotrakov replied:
Okay, let me try to add a `SerializationStrategy` static member to `SerializationUtils` and a new  `SerializationUtils.setSerializationStrategy()` method. Looks like it can make simplify the patch which is definitely good.

I am planning to get back to it on Nov 4th. I think we have time to finish it the next week. Thanks for the comments!

jgrandja commented:
@artem-smotrakov 

> We can also introduce a way how a user can set a list of allowed classes via environment variable or system property...

This can complicate things since we would need to specify a specific format so it can be read correctly from the environment. I'd rather keep things simple. Based on my previous comment on proposing to add `SerializationUtils.setSerializationStrategy()`, this will provide the ability for the application to read the allowed classes from the environment or system property or any source and than configure it with the proposed setter.

byference commented:
I found a problem in `JdkSerializationStrategy#ALLOWED_CLASSES` ，when I customize `UserDetails ` implementation (it is not in  `ALLOWED_CLASSES`  list) and use `RedisTokenStore` on `AuthorizationServerConfiguration#configure(AuthorizationServerEndpointsConfigurer)`，will throw `NotSerializableException`

~~~java
// org.springframework.security.oauth2.provider.token.store.redis.SaferObjectInputStream#resolveClass
protected Class<?> resolveClass(ObjectStreamClass classDesc) throws IOException, ClassNotFoundException {
    if (isProhibited(classDesc.getName())) {
        throw new NotSerializableException("Not allowed to deserialize " + classDesc.getName());
    }
    return super.resolveClass(classDesc);
}

private boolean isProhibited(String className) {
    for (String allowedClass : this.allowedClasses) {
        if (className.startsWith(allowedClass)) {
            return false;
        }
    }
    return true;
}
~~~


my `spring-security-oauth2 ` is ` 2.3.7.RELEASE` .

artem-smotrakov commented:
@jgrandja I have addressed your comments. Please take a look.

artem-smotrakov commented:
@vaddemgen @byference 

You can now implement a custom `SerializationStrategy` which allows deserializing classes which you need. For example, you can extend `WhitelistedSerializationStrategy`:

```java
public class CustomSerializationStrategy extends WhitelistedSerializationStrategy {

    private static final List<String> ALLOWED_CLASSES = new ArrayList<>();
    static {
        ALLOWED_CLASSES.add("java.lang.");
        ALLOWED_CLASSES.add("java.util.");
        ALLOWED_CLASSES.add("org.springframework.security.");
        ALLOWED_CLASSES.add("your.custom.impl.");
    }

    public CustomSerializationStrategy() {
        super(ALLOWED_CLASSES);
    }
}
```

Then, you can set your serialization strategy with `SerializationUtils.setSerializationStrategy()` method. Or, you can define it in the `META-INF/spring.factories` file:

```
org.springframework.security.oauth2.common.util.SerializationStrategy = custom.impl.CustomSerializationStrategy
```

I tested it with a demo application, it seems to work. Could you please try it out as well?

byference commented:
`SerializationUtils.setSerializationStrategy()`  and `org.springframework.security.oauth2.common.util.SerializationStrategy` is not available in the latest version (`2.3.7.RELEASE `) . 

now, I can customizing the implementation of `RedisTokenStoreSerializationStrategy` （refer to`JdkSerializationStrategy`）to solve my problem,

~~~java
public class CustomJdkSerializationStrategy extends StandardStringSerializationStrategy {
    ...
    static {
        List<String> classes = new ArrayList<String>();
        classes.add("java.lang.");
        classes.add("java.util.");
        classes.add("org.springframework.security.");
        classes.add("my.custom.impl.");
        ALLOWED_CLASSES = Collections.unmodifiableList(classes);
    }

    @Override
    @SuppressWarnings("unchecked")
    protected <T> T deserializeInternal(byte[] bytes, Class<T> clazz) {
        ...
        try {
            SaferObjectInputStream saferObjectInputStream = new SaferObjectInputStream(new ByteArrayInputStream(bytes), ALLOWED_CLASSES);
            return (T) saferObjectInputStream.readObject();
        } catch (Exception e) {
            throw new SerializationFailedException("Failed to deserialize payload", e);
        }
    }
    ...
    // copied from org.springframework.security.oauth2.provider.token.store.redis.SaferObjectInputStream
    class SaferObjectInputStream extends ObjectInputStream {
        ...
    }
}


@Configuration
@EnableAuthorizationServer
protected static class AuthorizationServerConfiguration extends AuthorizationServerConfigurerAdapter {
    ...
    @Override
    public void configure(AuthorizationServerEndpointsConfigurer endpoints) throws Exception {
        RedisTokenStore redisTokenStore = new RedisTokenStore(redisConnectionFactory);
        redisTokenStore.setSerializationStrategy(new CustomJdkSerializationStrategy());
        endpoints.tokenStore(redisTokenStore)
            .authenticationManager(authenticationManager);
    }
}
~~~

but i don't think it's a good way, so looking forward to the new version @artem-smotrakov

artem-smotrakov commented:
@byference The issue you are facing is #1783. But this is a pull request which addresses this issue. The pull request is targeted to `2.4.0`.

artem-smotrakov commented:
@jgrandja 

> Also, can you add a test for `JdbcClientTokenServices`

I added a couple of test cases for `JdbcClientTokenServicesTests`.

artem-smotrakov commented on spring-security-oauth2/src/test/java/org/springframework/security/oauth2/common/util/SerializationUtilsTest.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * @author Artem Smotrakov
+ */
+public class SerializationUtilsTest {
```
@jgrandja I added `CustomSerializationStrategyTest` which mocks `SpringFactoriesLoader.loadFactories()`. I borrowed the idea from 

https://github.com/spring-projects/spring-security/blob/00f8991facc6b615d6e2824c9b9ceff7b415d57e/config/src/test/java/org/springframework/security/config/annotation/web/WebSecurityConfigurerAdapterPowermockTests.java

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SerializationUtils.java:
```diff
@@ -1,64 +1,73 @@
+    /**
+     * @return The current serialization strategy.
+     */
+    public static SerializationStrategy getSerializationStrategy() {
```
Ok we can leave as-is.

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/WhitelistedSerializationStrategy.java:
```diff
@@ -0,0 +1,146 @@
+ * @author Artem Smotrakov
+ * @since 2.4
+ */
+public class WhitelistedSerializationStrategy implements SerializationStrategy {
```
> a user can extend this class to use it

Even if a user were to extend this class, there aren't any `protected` methods where a user can override the behaviour. They would only be able to override `serialize` and/or `deserialize` to change behaviour. Instead of extension, a user can implement a delegation-based approach to achieve the same.

> Do you have a reason why this class has to be final?

To prevent extension so we are not on the hook for support. However, given that 2.4.0 is the last minor version and this project will soon be deprecated, it's pretty much a moot point. I would be fine if we left this as-is.

jgrandja commented on spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/WhitelistedSerializationStrategy.java:
```diff
@@ -0,0 +1,162 @@
+        @Override
+        protected Class<?> resolveClass(ObjectStreamClass classDesc)
+                throws IOException, ClassNotFoundException {
+            if (isProhibited(classDesc.getName())) {
```
As it stands now, when an application upgrades to 2.4 and have extended `OAuth2AccessToken` (or `OAuth2RefreshToken` and `OAuth2Authentication`), it will be forced to configure a custom `WhitelistedSerializationStrategy`, otherwise the application will fail when attempting to deserialize.

The preference would be that the application will continue to work after upgrading to 2.4 without having to provide an additional configuration (code change). Can we add an additional check for superclass? For example, `classDesc.forClass().getSuperclass()`

artem-smotrakov replied:
@jgrandja 

> As it stands now, when an application upgrades to 2.4 and have extended `OAuth2AccessToken` (or `OAuth2RefreshToken`, `OAuth2Authentication`), it will be forced to configure a custom `WhitelistedSerializationStrategy`, otherwise the application will fail when attempting to deserialize.

Unfortunately yes, that's how it's going to be.

> The preference would be that the application will continue to work after upgrading to 2.4 without having to provide an additional configuration (code change). 

Yes, ideally.

> Can we add an additional check for superclass? For example, `classDesc.forClass().getSuperclass()`

I am not sure if we should do that. If we add such a check, an attacker can prepare a class `MaliciousOAuth2AccessToken` which extends `DefaultOAuth2AccessToken`, and put a malicious code to the constructor. Then, `MaliciousOAuth2AccessToken` can pass the `is Prohibited()` check.

jgrandja replied:
@artem-smotrakov 

> I am not sure if we should do that. If we add such a check, an attacker can prepare a class `MaliciousOAuth2AccessToken`...

Yes, you're right. Ok, so we should not apply check on superclass?

> As it stands now, when an application upgrades to 2.4 and have extended `OAuth2AccessToken` (or `OAuth2RefreshToken`, `OAuth2Authentication`), it will be forced to configure a custom `WhitelistedSerializationStrategy`, otherwise the application will fail when attempting to deserialize.

This change introduces incompatibility changes between 2.3.x and 2.4.0, which cannot happen in a minor  release. An application built against 2.3.x should be able to upgrade to 2.4.0 without any issues. This change would be fine for a major version though. Please see [Versioning](https://apr.apache.org/versioning.html#strategy) for additional details.

In order to get this change into 2.4.0, this will need to be an "opt-in" feature rather than the default configuration. So if the application wants stricter deserialization, than they would need to explicitly configure `WhitelistedSerializationStrategy`. For the default configuration, the application should perform the same as it did in 2.3.x.

artem-smotrakov replied:
@jgrandja 

> In order to get this change into 2.4.0, this will need to be an "opt-in" feature rather than the default configuration.

That's sad. I would prefer to make it a bit more secure by default. I'll update it to be an opt-it feature, however I don't like it.

jgrandja replied:
@artem-smotrakov 

> I would prefer to make it a bit more secure by default

I definitely would too. But with the way versioning works, this would break existing applications and users would log new issues around this as we have seen with the 2.3.7 release.

artem-smotrakov replied:
@jgrandja 

* I added a new `DefaultSerializationStrategy` which allows all classes by default. The `SerializationUtils` uses `DefaultSerializationStrategy` by default.
* I also mentioned in javadoc that the default settings may be unsafe if data come from an untrusted source.
* Updated tests

Please let me know what you think.

artem-smotrakov commented:
> Also, I would like to make you aware (so there are no surprises), that the main reason for releasing 2.4 is to officially `@Deprecated` all classes in `spring-security-oauth2`. This will also include the new classes introduced in this PR. However, the 2.4.x branch will be supported until approx. March 2021 so this improvement is still valuable until than. Keep an eye on the blog post next week (same time as release Nov 14) for further details.

Sure, thanks for letting me know!

artem-smotrakov commented on spring-security-oauth2/src/test/java/org/springframework/security/oauth2/client/token/JdbcClientTokenServicesTests.java:
```diff
@@ -79,4 +85,37 @@ public void testSaveAndRemoveToken() throws Exception {
+		SerializationStrategy oldStrategy = SerializationUtils.getSerializationStrategy();
+		try {
+			SerializationUtils.setSerializationStrategy(newStrategy);
+			OAuth2AccessToken accessToken = new DefaultOAuth2AccessToken("FOO");
```
Yes, good catch! I also forget to add a couple of checks in this test case.

jgrandja commented:
Thanks you @artem-smotrakov ! This is now in master.

artem-smotrakov commented:
> Thank you @artem-smotrakov ! This is now in master.

Awesome, thanks for the review @jgrandja !
